feat(caip): add caip-related utils/constants to caip package

gomesalexandre opened the pull request:
This moves the CAIP-related utils as well as related constants from web.

### Issue

closes #557

### Notes

This only moves utils that are strictly related to CAIPs.
If the input/output of such util is something from a domain-specific logic that's not *strictly* related to CAIPs (e.g account type, label, UTXO params etc...), it should not belong here and should stay defined where it is currently

0xdef1cafe commented on packages/caip/src/utils.ts:
```diff
@@ -0,0 +1,42 @@
+
+// TODO(gomes): Do we still want to call this "reference" or do we want a higher-level naming for this?
+// We're already using accountId, assetId, and chainId higher-level terminologies vs. the specc-y ones from CAIP
+export const getCaip2Reference = (caip2: CAIP2) =>
```
better yet - delete it, not used anywhere in web or lib

gomesalexandre replied:
It's not used anywhere because it doesn't exist yet, it's a newly introduced function 👀 , currently it is the `caip2Reference` var:
https://github.com/shapeshift/web/blob/60ffe3ffaa2e2198c72ad475dabbdcaa88d65819/src/lib/ens.ts#L21-L25

But to use it programmatically, we need to make it a function

0xdef1cafe replied:
@gomesalexandre in that case - let's use the new vernacular `chainId` over `caip2`

0xdef1cafe commented on packages/caip/src/utils.ts:
```diff
@@ -0,0 +1,42 @@
+
+// TODO(gomes): Do we still want to call this "reference" or do we want a higher-level naming for this?
+// We're already using accountId, assetId, and chainId higher-level terminologies vs. the specc-y ones from CAIP
+export const getReferenceFromChainId = (caip2: CAIP2) =>
```
i still have no idea what this is doing or why. what reference are we referring to? chain reference? we shouldn'y be using old terminology or types in new code either, make this param chainId etc

gomesalexandre replied:
Agreed, `chainId` makes more sense as an argument. Changed it and removed the comment as this effectively answers my question there.
As far as the naming of the util is concerned, I've named it getChainReferenceFromChainId which is still CAIP-specific.

Maybe something like `networkId` would be a good terminology to use as high-level equivalent of the chain reference?

0xdef1cafe commented on packages/caip/src/utils.ts:
```diff
@@ -0,0 +1,42 @@
+export const accountIdToSpecifier = (accountId: AccountSpecifier): string => {
+  // in the case of account based chains (eth), this is an address
+  // in the case of utxo based chains, this is an x/y/zpub
+  return accountId.split(':')[2] ?? ''
```
a lot of these implementations were turbo dirty hacks with no safety - we should call underlying safer methods over dumping things in here from web

0xdef1cafe reviewed (changes requested):
see comments plz

gomesalexandre commented on packages/caip/src/utils.ts:
```diff
@@ -0,0 +1,42 @@
+  [cosmosChainId]: 'cosmos:cosmoshub-4/slip44:118'
+}
+// TODO(gomes): this probably needs better terminology than "supported"?
+const supportedChainIds = [ethChainId, btcChainId, cosmosChainId] as const
```
Yeah it should live in a better place. I think `chainIdToAssetId` and the various chainIDs / assetIDs consts still make sense to keep here though.
One could argue that their use is ShapeShift specific, but what we're really doing here is doing some augmentation on top of caip data, that's actually domain-agnostic.

gomesalexandre commented:
### Improved namespacing and added const:

- Move CAIP2-related utils to `/caip2/` namespace in https://github.com/shapeshift/lib/pull/572/commits/f8a37f506f8c4909a0d2ebfec4fae121a9180a43 - improved the typing of the constants while at it
- Added `networkTypeToChainReference`, a mirror of `chainReferenceToNetworkType`, in https://github.com/shapeshift/lib/pull/572/commits/f8a37f506f8c4909a0d2ebfec4fae121a9180a43
  - `getChainReferenceFromChainId` is now getting `network` from `fromCaip2` and accessing the reference from that mirrored const. Still not sure about that implementation, we could possibly get `chain` and `network` from `fromCaip2` instead and just access `shapeShiftToCAIP2[chainType].references[network]`
 
### No more sneaky hacks:

- Update `accountIdToSpecifier` to return `account` from `fromCaip10` in https://github.com/shapeshift/lib/pull/572/commits/f8a37f506f8c4909a0d2ebfec4fae121a9180a43
- Update `assetIdToChainId` to use `fromCaip19` and `toCaip2` to make a chainid from an assetId in a1913ce504281c5e8476b30a43826cebb08d588e
- Update `accountIdToChainId` to use `fromCaip10` to make a chainId from an accountId in a1913ce504281c5e8476b30a43826cebb08d588e

gomesalexandre commented on packages/caip/src/caip19/caip19.ts:
```diff
@@ -53,24 +59,6 @@ const validAssetNamespaces = Object.freeze({
   ]
 })
 
-const chainReferenceToNetworkType: Record<string, NetworkTypes> = Object.freeze({
```
It's only used internally in `fromCAIP19`:
https://github.com/shapeshift/lib/blob/deca9e9dcee1933f6ef152b7a8d530613368d322/packages/caip/src/caip19/caip19.ts#L143

theoboldfrazier reviewed (approved):
Looks good!

0xdef1cafe commented on packages/caip/src/utils.test.ts:
```diff
@@ -0,0 +1,74 @@
+} from './utils'
+
+describe('accountIdToChainId', () => {
+  it('can get eth caip2 from accountId', () => {
```
```suggestion
  it('can get eth chainId from accountId', () => {
```

0xdef1cafe commented on packages/caip/src/utils.test.ts:
```diff
@@ -0,0 +1,74 @@
+    expect(chainId).toEqual(ethChainId)
+  })
+
+  it('can get btc caip2 from accountId', () => {
```
```suggestion
  it('can get btc chainId from accountId', () => {
```

0xdef1cafe commented on packages/caip/src/utils.test.ts:
```diff
@@ -0,0 +1,74 @@
+    expect(result).toEqual(address)
+  })
+
+  it('can get xpub form accountId', () => {
```
```suggestion
  it('can get xpub from accountId', () => {
```

0xdef1cafe commented on packages/caip/src/utils.test.ts:
```diff
@@ -0,0 +1,74 @@
+  })
+})
+
+describe('assetIdtoChainId', () => {
```
```suggestion
describe('assetIdToChainId', () => {
```

0xdef1cafe commented on packages/caip/src/utils.ts:
```diff
@@ -0,0 +1,40 @@
+  [btcChainId]: 'bip122:000000000019d6689c085ae165831e93/slip44:0',
+  [cosmosChainId]: 'cosmos:cosmoshub-4/slip44:118'
+}
+export const assetIdtoChainId = (assetId: AssetId): string => {
```
```suggestion
export const assetIdToChainId = (assetId: AssetId): string => {
```

0xdef1cafe commented on packages/caip/src/utils.ts:
```diff
@@ -0,0 +1,40 @@
+  return chainId
+}
+
+export const accountIdToChainId = (accountId: AccountId): ChainId => {
+  // accountId = 'eip155:1:0xdef1...cafe
+  const { caip2 } = fromCAIP10(accountId)
+  return caip2
+}
```
```suggestion
// accountId = 'eip155:1:0xdef1...cafe
export const accountIdToChainId = (accountId: AccountId): ChainId => fromCAIP10(accountId).caip2
```

0xdef1cafe commented on packages/caip/src/utils.ts:
```diff
@@ -0,0 +1,40 @@
+  return caip2
+}
+
+export const accountIdToSpecifier = (accountId: AccountId): string => {
+  const { account } = fromCAIP10(accountId)
+  return account
+}
```
```suggestion
export const accountIdToSpecifier = (accountId: AccountId): string => fromCAIP10(accountId).account
```

0xdef1cafe commented on packages/caip/src/utils.ts:
```diff
@@ -0,0 +1,40 @@
+  return account
+}
+
+export const getChainReferenceFromChainId = (chainId: ChainId) => {
+  const { network } = fromCAIP2(chainId)
+  const chainReference = networkTypeToChainReference[network]
+  return chainReference
+}
```
```suggestion
export const getChainReferenceFromChainId = (chainId: ChainId) => networkTypeToChainReference[fromCAIP2(chainId).network]
```

0xdef1cafe commented on packages/caip/src/utils.ts:
```diff
@@ -0,0 +1,40 @@
+}
+export const assetIdtoChainId = (assetId: AssetId): string => {
+  const { chain, network } = fromCAIP19(assetId)
+  const chainId = toCAIP2({ chain, network })
+
+  return chainId
```
```suggestion
  return toCAIP2({ chain, network })
```

0xdef1cafe commented on packages/caip/src/utils.test.ts:
```diff
@@ -0,0 +1,74 @@
+})
+
+describe('getChainReferenceFromChainId', () => {
+  it('returns the reference from a caip2 string with a valid reference', () => {
```
```suggestion
  it('returns the reference from a chainId string with a valid reference', () => {
```

0xdef1cafe commented on packages/caip/src/utils.ts:
```diff
@@ -0,0 +1,27 @@
+export const btcChainId = 'bip122:000000000019d6689c085ae165831e93'
+export const cosmosChainId = 'cosmos:cosmoshub-4'
+
+export const chainIdToAssetId: Record<string, string> = {
```
```suggestion
export const chainIdToAssetId: Record<ChainId, AssetId> = {
```
